how-to/workspace-platform-starter/client/src/modules/actions/dynamic-menu/actions.ts

```diff
@@ -0,0 +1,135 @@
+	 * @param loggerCreator For logging entries.
+	 * @param helpers Helper methods for the module to interact with the application core.
+	 * @returns Nothing.
+	 */
+	public async initialize(
+		definition: ModuleDefinition<DynamicMenuProviderOptions>,
+		loggerCreator: LoggerCreator,
+		helpers: ModuleHelpers
+	): Promise<void> {
+		this._logger = loggerCreator("DynamicMenuProvider");
+		this._helpers = helpers;
+	}
+
+	/**
+	 * Get the actions from the module.
+	 * @param platform The platform module.
+	 * @returns The map of custom actions.
+	 */
+	public async get(platform: WorkspacePlatformModule): Promise<CustomActionsMap> {
+		const actionMap: CustomActionsMap = {};
+
+		actionMap["pages-menu"] = async (payload: CustomActionPayload): Promise<void> => {
+			if (payload.callerType === CustomActionCallerType.CustomButton) {
+				const pages: Page[] = await platform.Storage.getPages();
+				const res = await this.showCustomMenu(
+					{ x: payload.x, y: payload.y },
+					payload.windowIdentity,
+					"There are no pages",
+					pages
+						.map((p) => ({
+							label: p.title,
+							id: p.pageId
+						}))
+						.sort((a, b) => a.label.localeCompare(b.label))
+				);
+
+				if (!isEmpty(res) && this._helpers?.launchPage) {
+					const pageToLaunch = await platform.Storage.getPage(res);
+					await this._helpers.launchPage(pageToLaunch);
+				}
+			}
+		};
+
+		return actionMap;
+	}
+
+	/**
+	 * Show a custom menu.
+	 * @param position The position to show the menu.
+	 * @param position.x The x position to show the menu.
+	 * @param position.y The y position to show the menu.
+	 * @param parentIdentity The identity of the parent window.
+	 * @param noEntryText The text to display if there are no entries.
+	 * @param menuEntries The menu entries to display.
+	 * @returns The menu entry.
+	 */
+	private async showCustomMenu(
+		position: { x: number; y: number },
+		parentIdentity: OpenFin.Identity,
+		noEntryText: string,
+		menuEntries: { label: string; id: string; icon?: string }[]
+	): Promise<string | undefined> {
+		const dockWindow = fin.Window.wrapSync(parentIdentity);
+		const dockBounds = await dockWindow.getBounds();
+
+		const platformWindow = fin.Window.wrapSync(fin.me.identity);
+
+		const menuPos = Math.floor((position.x + 20) / 40) * 40;
+
+		const currentPalette = await this._helpers?.getCurrentPalette();
+
+		const id = randomUUID();
+		const result = await platformWindow.showPopupWindow({
+			name: id,
+			initialOptions: {
+				showTaskbarIcon: false,
+				backgroundColor: currentPalette?.backgroundPrimary,
+				customData: {
+					noEntryText,
+					menuEntries,
+					palette: {
+						backgroundPrimary: currentPalette?.backgroundPrimary,
+						textDefault: currentPalette?.textDefault,
+						inputBackground: currentPalette?.inputBackground
+					}
+				}
+			},
+			url: "http://localhost:8080/common/popups/dynamic-menu/index.html",
```

Comment:
I'm guessing we have something that does localhost replacement in code? Have we got a note of that in any of the docs?

Reply:
I moved it to config